segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,382 @@
+        return CompletableFuture.supplyAsync(() -> syncExists(streamSegmentName, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return CompletableFuture.supplyAsync(() -> {
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            if (!Files.exists(path)) {
+                throw new CompletionException(new StreamSegmentNotExistsException(streamSegmentName));
+            } else if (Files.isWritable(path)) {
+                return FSSegmentHandle.getWriteHandle(streamSegmentName);
+            } else {
+                try {
+                    return openRead(streamSegmentName).get();
+                } catch (InterruptedException | ExecutionException e) {
+                    throw new CompletionException(e);
+                }
+            }
+        }, executor);
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return CompletableFuture.supplyAsync( () -> syncCreate(streamSegmentName, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return CompletableFuture.supplyAsync(() -> syncWrite(handle, offset, data, length, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return CompletableFuture.supplyAsync(() -> syncSeal(handle, timeout), executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return CompletableFuture.supplyAsync(() -> syncConcat(targetHandle, offset, sourceSegment, timeout),
+                    executor);
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return CompletableFuture.supplyAsync(() -> syncDelete(handle, timeout), executor);
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(streamSegmentName));
+        }
+
+        return FSSegmentHandle.getReadHandle(streamSegmentName);
+    }
+
+
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length, Duration
+            timeout) {
+
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (!Files.exists(path)) {
```

Comment:
We are discussing this in another thread. Will update this location according to the outcome of the other.